### ISSUE TITLE
Add a segment for GitHub notifications

### ISSUE THREAD

**Issue by stardiviner** [seagle0128/doom-modeline#66]
As like this config: https://github.com/stardiviner/emacs.d/blob/develop/init/Emacs/init-custom-mode-line.el#L946
Can you add a segment for GitHub notifications? Also provide an option to switch between all notifications or participated notifications.

I used my custom mode-line for a few years, Now I found doom-modeline, because it is fast. I switched to doom-modeline now. BTW, honestly, doom-themes does not great on my machine. Colors seems not very same with screenshot. Don't know why.

**Comment by seagle0128** [comment]
Thanks for liking `doom-modeline`!

I have to say `ghub+` is not a popular package although I am using it with `magit`. I need some time to  investigate and make the decision.

Regarding `doom-themes`, are you using Emacs in terminal or with GUI? That's different. If you are using Emacs GUI, please refer to my config: https://github.com/seagle0128/.emacs.d/blob/6018a1592457b2f3c77961149a7c0b14cbe88991/lisp/init-ui.el#L98 .

**Comment by seagle0128** [comment]
Following by the snippets you provide, I create a new segment of Github notifications. But honestly I couldn't test a full functionality.  I don't know if `ghubp-get-notifications` works well in the timer. Please trial and help to test. Thanks in advance!

Screenthot

![github](https://user-images.githubusercontent.com/140797/49808406-6b7c5700-fd97-11e8-83bb-d2ddbe5be8eb.png)

Customization
```elisp
;; Whether display github notifications or not.
(setq doom-modeline-github t)

;; The interval of checking github.
(setq doom-modeline-github-interval (* 30 60))
```

**Comment by seagle0128** [comment]
I did some tests and got the notifications number from github, but disappear a few minutes. `ghubp-get-notifications` returns errors after the first success. I have no idea if it's as designed, but it seems work as expected in mode-line. The performance is not good. It cost huge CPU while fetching from github, but I think it's fine since it's triggered every 30 min.

I am waiting for more feedbacks.

**Comment by stardiviner** [comment]
Sorry for late response, I cloned repo for testing latest commits. Need wait for participation notifications on my GitHub account to test. So feedback maybe later.

About the doom-themes, I applied following config:
```elisp
(use-package doom-themes
  :ensure t
  :no-require t
  :config (load-theme 'doom-one t)
  ;; Enable flashing mode-line on errors
  (doom-themes-visual-bell-config)

  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (doom-themes-treemacs-config)

  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))

(use-package solaire-mode
  :ensure t
  :hook (((change-major-mode after-revert ediff-prepare-buffer) . turn-on-solaire-mode)
         (minibuffer-setup . solaire-mode-in-minibuffer)
         (after-load-theme . solaire-mode-swap-bg)))
```

Here is my screenshot:

![image](https://user-images.githubusercontent.com/378699/49840646-44517400-fdef-11e8-81b4-6883e9f0c4f2.png)

**Comment by stardiviner** [comment]
And I found the flycheck segment is partially hidden.

As in this screenshot:

![image](https://user-images.githubusercontent.com/378699/49840962-a2cb2200-fdf0-11e8-9af0-c2260b2e9e31.png)

**Comment by seagle0128** [comment]
The new screenshot above looks as good as mine, except the icons of `treemacs`.

![image](https://user-images.githubusercontent.com/140797/49849884-6f02f300-fe16-11e8-9e49-2b3ff040aaa4.png)

With ` (doom-themes-treemacs-config)`, the icons will be changed to atom style.

About the `flycheck` segment issue, I revert my recent change for it. It works well on macOS, but seems not on Linux. Please check the latest version.

BTW, I can open a new issue in your repo for testing the notification feature 😄

**Comment by stardiviner** [comment]
I guess the slight color difference is because system reason. seems macOS has better rendering. Close this issue now.

**Comment by seagle0128** [comment]
@TxGVNN It depends. If the feature brings big performance impact, I won't. And I am always trying to add enablers for the segments. You can choose to enable or disable. So far it's still fast according my test.

**Comment by seagle0128** [comment]
@stardiviner My only concern is if (ghubp-get-notifications :participating "true") is executed correctly. When I was evaluating manually, it returned errors sometimes. Does it need auth or token? I just leverage your snippets to implement.

And I am thinking if `run-with-idle-timer` brings less impacts while fetching the data. Your thoughts?

**Comment by stardiviner** [comment]
Seems the segment has lot of reduntant running process to access the API.
Here is the screenshot:

![image](https://user-images.githubusercontent.com/378699/49976000-3f213000-ff7b-11e8-977a-d03b7e4af480.png)

**Comment by seagle0128** [comment]
@stardiviner Are all processes alive?

**Comment by stardiviner** [comment]
Yes, all processes alive, and they are from `list-process`. I'm using Arch Linux.

**Comment by stardiviner** [comment]
Recently, I checked out this multiple connections issue, I found might because that time my network not avialbe, so that multiple connections not closed. Anyway, now it works fine. Thanks @seagle0128

**Comment by seagle0128** [comment]
Welcome. Glad to know! Thanks for the update.

**Comment by seagle0128** [comment]
@stardiviner I reimplemented this feature with `async` and `ghub`. If you are using `magit`, all packages are ready. `ghub+` can be dropped now. Moreover, the `checking` is asynchronous so won't block the typing. Better experience!

**Comment by stardiviner** [comment]
@seagle0128 Wow! That's awesome, thanks. :)

**Comment by edkolev** [comment]
I don't want the github segment - would it be possible to run this code *only if* the github segment is in the modeline?

```
(run-with-timer 30
                doom-modeline-github-interval
                #'doom-modeline--github-fetch-notifications)
```

I can set `(setq doom-modeline-github nil)`, however, the above snippet will still be executed.

**Comment by seagle0128** [comment]
@edkolev Yes, it's possible. The fetching action in the anonymous timer is executed every 30 minutes. If `doom-modeline-github` is nil, nothing to do and then exits. If just starting the timer when `doom-modeline-github` is t, you have to restart Emacs after change `doom-modeline-github`. Do you have any better idea?

**Comment by seagle0128** [comment]
@edkolev I have an idea and submitted the changes. But only support the runtime change of `doom-modeline-github` on 26+. On 25, Emacs need to be restarted to take effect after `doom-modeline-github` is changed.

**Comment by edkolev** [comment]
Thanks, much better - no code will be executed when github segment is disabled!

However, `add-variable-watcher` seems like an overkill, I think it's perfectly fine for the user to restart emacs when changing this variable.

**Comment by seagle0128** [comment]
Thanks for the feedback and comments. I will keep `add-variable-watcher` for debugging for now.
